### PR TITLE
EL-1540 use default values for feature flags

### PR DESCRIPTION
**Pull request by willc-work**
[Jira ticket](https://dsdmoj.atlassian.net/browse/EL-1540)

When a check is created without a particular session feature flag we now use the default for that flag.
Moved some methods out of the enabled? method as the session_flags method was also using `enabled?` but didnt need all of the functionality. So 'overrideable_flag' and 'env_var_flag' were extracted to be used by both `session_flags` and `enabled?`
This allows us to revert to using the ENV_VAR to set the sessions flags at the start of the check, and also in the tests where we turn the flags on and off around specific tests. Elsewhere it will use the default for the flag.

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing
- Branch is generally up to date with main Github - definitely no conflicts
- No unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- PR description says *what* changed and *why*, with a link to the JIRA story.
- Diff has been checked for unexpected changes being included.
- Commit messages say why the change was made.

### CODE REVIEW

**Review by HettieS — APPROVED** [review]
I have approved with a comment to rename a test. I found the new complexity quite hard to follow - can we please add some information to the README on the purpose of the new default setting? I would quite like that to include a hierarchy of where the flag value will come from (or a few written use cases in the README to display it clearly)

File: spec/models/feature_flags_spec.rb [hunk 1]
```diff
@@ -69,6 +69,34 @@
...
+        ENV["EXAMPLE_FEATURE_FLAG"] = "enabled"
+        each.run
+        ENV["EXAMPLE_FEATURE_FLAG"] = "disabled"
+      end
+
+      it "uses the ENV_VAR to set the feature flag" do
+        expect(described_class.session_flags.fetch("example")).to eq true
+      end
+
+      context "with ff" do
```
[the remark this answers is not in the document]
Comment (willc-work) [reply spec/models/feature_flags_spec.rb:84]: oh sorry about this, it was a placeholder until we worked out what the test should be doing. Will fix.

File: app/lib/feature_flags.rb [hunk 2]
```diff
@@ -22,19 +27,24 @@ def enabled?(flag, session_data = nil, without_session_data: false)
...
 
       if STATIC_FLAGS.key?(flag)
-        ENV["#{flag.to_s.upcase}_FEATURE_FLAG"]&.casecmp("enabled")&.zero? || STATIC_FLAGS.fetch(flag).fetch(:default)
+        # if the flag is a valid flag and of type `session` but does not
+        # exist in the session_data we return the default value for that flag
+        if STATIC_FLAGS[flag][:type] == "session"
+          STATIC_FLAGS[flag][:default]
+        else
+          # if it is a global flag we check the env_var
+          # if there is no env_var specified we use the default value for that flag
```
Comment (HettieS) [inline app/lib/feature_flags.rb:45]: I'm not sure if this is a bit scary or not. Why would we use the default value if there is no env_var? Could we risk accidentally turning something on?

Comment (HettieS) [reply app/lib/feature_flags.rb:45]: In fact maybe I am misunderstanding here. Is this setting purely for testing? Where we might want to have a default but there is no env var? I guess my fear is accidentally leaving a `default: true` and forgetting to set an env_var in prod. Maybe I am overthinking it. Maybe a comment would help here.

File: app/lib/feature_flags.rb [hunk 3]
```diff
@@ -22,19 +27,24 @@ def enabled?(flag, session_data = nil, without_session_data: false)
...
-      end
+      possible_override = overrideable_flag(flag)
+      return possible_override unless possible_override.nil?
 
       if STATIC_FLAGS.key?(flag)
-        ENV["#{flag.to_s.upcase}_FEATURE_FLAG"]&.casecmp("enabled")&.zero? || STATIC_FLAGS.fetch(flag).fetch(:default)
+        # if the flag is a valid flag and of type `session` but does not
+        # exist in the session_data we return the default value for that flag
+        if STATIC_FLAGS[flag][:type] == "session"
+          STATIC_FLAGS[flag][:default]
```
Comment (HettieS) [inline app/lib/feature_flags.rb:42]: so do we never look at the env_var anymore for session flags (when it is not in session_data/there is no session_data)? We only ever use the default?

Comment (HettieS) [reply app/lib/feature_flags.rb:42]: How would we then manage session flags being switched on or off across multiple envs? Perhaps I am misunderstanding.

File: app/lib/feature_flags.rb (hunk 2 above)
Comment (willc-work) [reply app/lib/feature_flags.rb:45]: Yes this is only ever expected to fire during testing when we provide an env_var in the before block and no session_data.
The defaults are also for tests only, they are added so that when a flag doesn't have a value we give it a value that will not break our tests.
I will add more to the comment to explain the env_var in tests.

File: app/lib/feature_flags.rb (hunk 3 above)
Comment (willc-work) [reply app/lib/feature_flags.rb:42]: >never look at the env_var anymore for session flags
correct. If a check starts without a new feature and we enable the env_var to release the feature, we would be passing 'enabled' to the check, which would then break.
so the only time we should check the env_var (for session flags) is when the check starts using the method `session_flags`

Comment (willc-work) [reply app/lib/feature_flags.rb:42]: Setting env_vars across multiple envs, or even mutiple UATs will work because the env_var is used to set the initial state of the check being undertaken. These flags are then set on a per check basis and not a per environment depending on the configuration of the deployment at the time the check started, so we can change env_vars between environments or even between checks on the same environment, and the check will use the inital configuration as its flags.

Comment (HettieS) [reply app/lib/feature_flags.rb:42]: Yeah I think I'm overthinking it. I was thinking, is there a situation where a session flag might not have session data passed in...(where we set `without_session_data: true`) but tbh I don't _think_ those should/would be session flags. 

I'm thinking for places like the start page where we don't have a session. But I wonder if we might have grouped a bunch changes together under a session flag, because they're all part of the same feature, but not pass the session in, in places where we don't have it. Probably overthinking it but we will just need to be a bit careful with our flag setup...

Comment (HettieS) [reply app/lib/feature_flags.rb:42]: It feels like a very edgy case nasty bug lol

File: app/lib/feature_flags.rb (hunk 2 above)
Comment (willc-work) [reply app/lib/feature_flags.rb:46]: i added a description as to what the default does rather than rename the attribute
